core/app/models/spree/address.rb

```diff
@@ -157,11 +157,17 @@ def country_iso=(iso)
 
     private
       def state_validate
-        # Skip state validation without country (also required)
-        # or when disabled by preference
-        return if country.blank? || !Spree::Config[:address_requires_state]
-        return unless country.states_required
+        # Skip state validation without country
+        return if country.blank?
 
+        # If states are not required we still check if state belongs to country
+        # If it doesn't we just remove it without adding state errors
+        unless country.states_required || readonly?
```

Comment:
@magnusvk Agreed. #immutable_merge should look at `country.states_required` to potentially leave the state_id out of the new_address